sequencer/src/block/full_payload/payload.rs

```diff
@@ -0,0 +1,233 @@
+use serde::{Deserialize, Serialize};
+use sha2::Digest;
+use std::{collections::HashMap, fmt::Display, sync::Arc};
+
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct Payload {
+    // Concatenated payload bytes for each namespace
+    #[serde(with = "base64_bytes")]
+    payload: Vec<u8>,
+
+    ns_table: NsTable,
+}
+
+impl Payload {
+    pub fn ns_table(&self) -> &NsTable {
+        &self.ns_table
+    }
+
+    /// Like [`QueryablePayload::transaction_with_proof`] except without the
+    /// proof.
+    pub fn transaction(&self, index: &Index) -> Option<Transaction> {
+        if !self.ns_table.in_bounds(index.ns()) {
+            return None;
+        }
+        let ns_id = self.ns_table.read_ns_id(index.ns());
+        let ns_payload = self.ns_payload(index.ns());
+        ns_payload.export_tx(&ns_id, index.tx())
+    }
+
+    // CRATE-VISIBLE HELPERS START HERE
+
+    pub(in crate::block) fn read_ns_payload(&self, range: &NsPayloadRange) -> &NsPayload {
+        NsPayload::from_bytes_slice(&self.payload[range.as_block_range()])
+    }
+
+    /// Convenience wrapper for [`Self::read_ns_payload`].
+    ///
+    /// `index` is not checked. Use `self.ns_table().in_bounds()` as needed.
+    pub(in crate::block) fn ns_payload(&self, index: &NsIndex) -> &NsPayload {
+        let ns_payload_range = self.ns_table().ns_range(index, &self.byte_len());
+        self.read_ns_payload(&ns_payload_range)
+    }
+
+    pub(in crate::block) fn byte_len(&self) -> PayloadByteLen {
+        PayloadByteLen(self.payload.len())
+    }
+}
+
+impl BlockPayload for Payload {
+    type Error = crate::Error;
+    type Transaction = Transaction;
+    type Instance = NodeState;
+    type Metadata = NsTable;
+
+    // TODO(BlockPayload): return type should not include `Self::Metadata`
+    fn from_transactions(
+        transactions: impl IntoIterator<Item = Self::Transaction>,
+        instance_state: &Self::Instance,
+    ) -> Result<(Self, Self::Metadata), Self::Error> {
+        // accounting for block byte length limit
+        let max_block_byte_len: usize = instance_state
+            .chain_config
+            .max_block_size
+            .try_into()
+            .map_err(|_| Self::Error::BlockBuilding)?;
+        let mut block_byte_len = NsTableBuilder::fixed_overhead_byte_len();
+
+        // add each tx to its namespace
+        let mut ns_builders = HashMap::<NamespaceId, NsPayloadBuilder>::new();
+        for tx in transactions.into_iter() {
+            // accounting for block byte length limit
+            block_byte_len += tx.payload().len() + NsPayloadBuilder::tx_overhead_byte_len();
+            if !ns_builders.contains_key(&tx.namespace()) {
+                // each new namespace adds overhead
+                block_byte_len += NsTableBuilder::ns_overhead_byte_len()
+                    + NsPayloadBuilder::fixed_overhead_byte_len();
+            }
+            if block_byte_len > max_block_byte_len {
```

Comment:
Why not throw an error if this happens?

Reply:
Previously we did not do anything (post log or return error) in this event. See #1382. If we throw an error here then we are breaking previous behaviour.

I put a `warn!` log because the user probably does not intend to trigger max block size, but strictly speaking it's not an "error".

Current enforcement of max block size is very primitive. I imagine in the future builders might want more fine-grained info about how each transaction contributes to block size. (Perhaps `from_transactions` might return more info instead of silently truncating the block!)

Anyway, I suggest any improvement here is a matter for a future PR. I suspect we do not wish to allocate resources to this task at the moment.

Reply:
Sounds good, thanks for the explanation.